Add technical docs on diagnosing instability etc (#85074)
elastic/elasticsearch: docs/reference/modules/discovery/fault-detection.asciidoc

similar-sounding messages on all other nodes like this:

```
[2020-01-29T10:20:48,985][INFO ][o.e.c.s.ClusterApplierService]
    [es-master-1] removed {
        {es-master-2}{1vSx_i__QCOjVyfA9kYEkg}{q78rb3MsRDyCcweeygxeiA}{10.244.35.101}{10.244.35.101:9300}{m}{xpack.installed=true},
        {es-data2-5}{KZMiSjy3QRiJevGQokB-zQ}{stIE-HcITQqLuCbAEgbCiA}{10.244.2.213}{10.244.2.213:9300}{d}{xpack.installed=true}},
    term: 469, version: 446676, reason: Publication{term=469, version=446676}
```

{es} emits this message on every node when applying a cluster state that
removes one or more nodes from the cluster. There's less useful information
here, and it's better to focus on the one from the `MasterService` which is
only emitted on the elected master. If you don't see the messages from the
`MasterService` then check you're looking at the logs from the elected master
covering the correct time period and that logging is enabled at `INFO` level.

Nodes will also log a message containing the string `master node changed`
whenever they start or stop following the elected master. You can use these
messages to determine each node's view of the state of the master over time.

> If a node restarts then it will leave the cluster and then join the cluster

Reviewer: ```suggestion
If a node restarts, it will leave the cluster and then join the cluster
```

again. When it rejoins, the `MasterService` will log that it is processing a
`node-join` task. You can tell from the master logs that the node was restarted
because the `node-join` message will indicate that it is

> `joining after restart`. In older versions you can also determine that a node

Reviewer: Do we need to include info about older versions? If so, we may want to note the actual version date.
```suggestion
`joining after restart`. In older {es} versions, you can also determine that a node
```

Author: I think so at least for now - I'm not going to backport these docs to unmaintained branches but I think they'll still be used with older versions.

restarted by looking at the second "ephemeral" ID in the `node-left` and
subsequent `node-join` messages: this ID is different each time the node starts
up. If a node is unexpectedly restarting then you'll need to look at the node's
own logs to see why it is shutting down.

If the node did not restart then you should look at the reason for its
departure in the `node-left` message, which is reported after each node. There

> are three possible reasons:

Author: I find it's pretty common for folks to miss this, I'd rather spell it out.

* `disconnected`: the connection from the master node to the removed node was
closed.

* `lagging`: the master published a cluster state update but the removed node
did not apply it within the permitted timeout. By default this timeout is 2

> minutes.

Author: This one is more questionable. It's not exactly recommended to change it but it might make sense in some cases. I'll link to the docs about those settings.

. The follower's `TransportService` running on thread
`elasticsearch[follower][transport_worker][T#3]` reads the incoming packets and
reconstructs and processes the check request. Usually the check succeeds
straight away and the same thread immediately constructs a response and passes
it back to the operating system.

. If, rarely, the check didn't succeed straight away (e.g. an election started
recently) then:

.. The follower's `FollowerChecker` running on thread
`elasticsearch[follower][cluster_coordination][T#4]` processes the request,
constructs a response, and tells the `TransportService` to send the response
back to the master.

.. The follower's `TransportService` running on thread
`elasticsearch[follower][transport_worker][T#3]` passes the response on to the
operating system.

. The operating system on the follower converts the response into one or more
packets and sends them out over the wire.

. Miscellaneous routers/firewalls/other devices between master & follower
forward the packets, possibly fragmenting or defragmenting them on the way.

. The operating system on the master receives the packets and notifies {es}
that they've been received.

> . The master's `TransportService` running on thread
> `elasticsearch[master][transport_worker][T#2]` reads the incoming packets,
> reconstructs the check response and processes it as long as the check didn't
> already time out.

Reviewer: ```suggestion
. The master's `TransportService`, running on thread
`elasticsearch[master][transport_worker][T#2]`, reads the incoming packets,
reconstructs the check response, and processes it as long as the check didn't
already time out.
```

> This means that there's a lot of different things that can delay the completion
> of a check for long enough for it to time out. Here are some examples for each
> step:

Reviewer: ```suggestion
There are a lot of different things that can delay the completion of a check and
cause it to time out. Here are some examples for each step:
```

. There may be a long garbage collection (GC) or virtual machine (VM) pause
after passing the check request to the `TransportService`.

. There may be a long wait for the specific `transport_worker` thread to become
available, or there may be a long GC or VM pause before passing the check
request onto the operating system.

> . A system fault (e.g. broken network card) on the master may delay sending the
> message over the wire, possibly indefinitely.

Reviewer: ```suggestion
. A system fault (for example, a broken network card) on the master may delay sending the
message over the network, possibly indefinitely.
```

> . Intermediate devices may delay, drop, or corrupt packets along the way. The
> OS on the master will wait and retransmit any unacknowledged or corrupted
> packets up to `net.ipv4.tcp_retries2` times; we recommend
> <<system-config-tcpretries,reducing this value>> since the default represents a
> very long delay.

Reviewer: ```suggestion
. Intermediate devices may delay, drop, or corrupt packets along the way. The
operating system for the master will wait and retransmit any unacknowledged or
corrupted packets up to `net.ipv4.tcp_retries2` times. We recommend
<<system-config-tcpretries,reducing this value>> since the default represents a
very long delay.
```

> . A system fault (e.g. broken network card) on the follower may delay receiving
> the message from the wire.

Reviewer: ```suggestion
. A system fault (for example, a broken network card) on the follower may delay receiving
the message from the network.
```

. There may be a long wait for the specific `transport_worker` thread to become
available, or there may be a long GC or VM pause during the processing of the
request on the follower.

. There may be a long wait for the `cluster_coordination` thread to become
available, or for the specific `transport_worker` thread to become available
again. There may also be a long GC or VM pause during the processing of the
request.

> . A system fault (e.g. broken network card) on the follower may delay sending
> the response over the wire.

Reviewer: ```suggestion
. A system fault (for example, a broken network card) on the follower may delay sending
the response from the network.
```

. Intermediate devices may delay, drop, or corrupt packets along the way again,
causing retransmissions.

> . A system fault (e.g. broken network card) on the master may delay receiving
> the message from the wire.

Reviewer: ```suggestion
. A system fault (for example, a broken network card) on the master may delay receiving
the message from the network.
```

. There may be a long wait for the specific `transport_worker` thread to become
available to process the response, or a long GC or VM pause.

To determine why follower checks are timing out, we can narrow down the reason
for the delay as follows:

> * Garbage collection (GC) pauses are recorded in the GC logs that {es} emits by

Reviewer: We spell out "garbage collection" above.
```suggestion
* GC pauses are recorded in the GC logs that {es} emits by
```

> default, and also usually by the `JvmMonitorService` in the main node logs, so
> confirm whether or not GC is resulting in delays.

Reviewer: ```suggestion
default, and also usually by the `JvmMonitorService` in the main node logs. Use
these logs to confirm whether or not GC is resulting in delays.
```

> * Virtual machine (VM) pauses also affect other processes on the same host. A
> VM pause also typically causes a discontinuity in the system clock, which {es}
> will report in its logs.

Reviewer: ```suggestion
* VM pauses also affect other processes on the same host. A
VM pause also typically causes a discontinuity in the system clock, which {es}
will report in its logs.
```

* Packet captures will reveal system-level and network-level faults, especially
if you capture the network traffic simultaneously at the elected master and the
faulty node. The connection used for follower checks is not used for any other
traffic so it can be easily identified from the flow pattern alone, even if TLS
is in use: almost exactly every second there will be a few hundred bytes sent
each way, first the request by the master and then the response by the
follower. You should be able to observe any retransmissions, packet loss, or
other delays on such a connection.

* Long waits for particular threads to be available can be identified by taking

> stack dumps (e.g. using `jstack`) or a profiling trace (e.g. using Java Flight

Reviewer: ```suggestion
stack dumps (for example, using `jstack`) or a profiling trace (for example, using Java Flight
```

Recorder) in the few seconds leading up to a node departure. The
<<cluster-nodes-hot-threads>> API sometimes yields useful information, but bear
in mind that this API also requires a number of `transport_worker` and

> `generic` threads across all the nodes in the cluster, so might be affected by

Reviewer: ```suggestion
`generic` threads across all the nodes in the cluster. The API may be affected by
```
